### PR TITLE
refactor(core): validate the whole weight load before the first upload

### PR DESCRIPTION
**Pull request by FeathBow**
The staged weight loader ran validate -> allocate -> upload serially per tensor, so a shape mismatch deep in the checkpoint surfaced only after earlier tensors were already written to the GPU. This splits the load into a recorded plan and its execution: entry methods (matrices, fused rows, column shards, and the norm vectors alike) validate against the config-derived shapes, allocate, and record the upload; `finish()` executes every recorded upload afterwards, so all load-boundary validation completes before the first H2D write. `finish` stays poisoned until every upload lands — a failed call cannot be retried into a vacuous success — and ownership is restructured so the loader holds every destination until `finish()` succeeds and `take()` redeems it: matrices can no longer be observed half-written on an error path, and the internal unsafe stays local to `finish` (sources are checker-borrowed from the shards, destinations are slot-owned). One log line reports the phase walls, e.g. `weight load: 398 uploads, alloc 279ms, upload 594ms`.

This is a fail-fast and observability refactor, not an optimization. 

## Test Env

Measured on one sm_89 GPU, Qwen3-4B, warm page cache, five repetitions per arm with means, arms paired in the same session (the surrounding startup phases drift with GPU clock state across sessions, so only same-session pairs are comparable):

| variant | warm load phase | HTTP-ready |
|---|---|---|
| base (per-tensor validate/alloc/upload) | 964 ms | 5126 ms |
| this change (validate everything, then upload) | 986 ms | 5128 ms |

The +22 ms is the price of deferring every upload behind full validation; HTTP-ready is unchanged within noise. Three performance variants were measured on the same rig and rejected:

- Running the uploads on a worker thread overlapped with construction reaches 905 ms, but the 36 ms over base buys a thread, raw cross-thread pointers, and a weaker validation-before-write ordering — not worth it for a correctness-sensitive path.
- Pre-growing the stream-ordered allocator's pool changes nothing: pool growth costs ~36 ms/GB regardless of call count (one 8 GB warm-up allocation costs the same ~290 ms as the 146 per-tensor ones), so the allocation wall is a per-byte floor, not per-call overhead.
- Routing the ~150 small vectors through the pinned stager costs +14% load phase (per-chunk event and thread overhead dwarfs a few-KB copy); they defer with the matrices but upload as plain pageable copies, which return only after the source is staged.

Validation on the same GPU, on the rebased tree: `hf_golden_gate` (logits through the new loader still match the HF golden), `sampling_behavior`, and the core unit tests pass; fmt and clippy clean.

### CODE REVIEW

File: openinfer-core/src/weight_loader.rs [hunk 1]
```diff
@@ -262,6 +268,319 @@ fn find_tensor<'a>(
...
+                },
+                PendingUpload::ColShard { src, plan } => unsafe {
+                    self.stager.upload_cols_at(src, &plan)?;
+                },
+                // SAFETY: destination validated at record time; a pageable
+                // async copy returns only after the source is staged, so the
+                // Cow may drop right after.
+                PendingUpload::Vector { src, dst_at } => unsafe {
+                    memcpy_htod_async(dst_at, bf16_bytes(&src), self.ctx.stream.cu_stream())
+                        .map_err(|e| anyhow::anyhow!("vector upload failed: {e}"))?;
```
[the remark this answers is not in the document]
Comment (FeathBow) [reply openinfer-core/src/weight_loader.rs]: Great. Done

**Comment by FeathBow** [comment]
Rebased onto main after #749 merged, and re-ran the paired A/B so the numbers match the current base.

The earlier figures in this PR were measured before #749 landed, when HTTP-ready was about 5.1 s. With #749 in main the ready baseline is about 3.6 s, so the absolute numbers needed refreshing even though the conclusion did not change.

One sm_89 GPU, Qwen3-4B, warm page cache, both arms in the same session, warm-up plus five repetitions each, means:

| arm | warm load phase | HTTP-ready |
|---|---|---|
| base | 934 ms | 3589 ms |
| this PR | 938 ms | 3585 ms |

The load phase is +4 ms (+0.4%) and HTTP-ready is unchanged (-5 ms). Per-run spread is 929-940 ms for base and 918-955 ms for the branch, so the load-phase delta sits inside run-to-run variation on this host. An earlier paired run on the pre-#749 base measured +22 ms (+2.3%); both runs agree the cost is small and that it does not reach HTTP-ready, but the delta itself is at the edge of what these measurements resolve.

The rebase had one conflict worth flagging: #749 changed the munmap in `from_safetensors_with_runtime` to a background thread, while this PR moves that function into `weights/load.rs`. Both changes are kept — the function lives in the new module and carries #749's async unmap verbatim.

Gates on the rebased tree, same host: fmt, clippy `-D warnings`, core unit tests, both GPU loader tests, qwen3 unit tests, `hf_golden_gate`, and `sampling_behavior` all pass.
